expression/builtin_miscellaneous.go

```diff
@@ -1046,3 +1049,79 @@ type uuidShortFunctionClass struct {
 func (c *uuidShortFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
 	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "UUID_SHORT")
 }
+
+type vitessHashFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *vitessHashFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTp := args[0].GetType().EvalType()
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp)
+	if err != nil {
+		return nil, err
+	}
+
+	bf.tp.Flen = 8
+	bf.tp.Flag |= mysql.UnsignedFlag
+	types.SetBinChsClnFlag(bf.tp)
+
+	sig := &builtinVitessHashSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_VitessHash)
+	return sig, nil
+}
+
+type builtinVitessHashSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinVitessHashSig) Clone() builtinFunc {
+	newSig := &builtinVitessHashSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalString evals VITESS_HASH(int64|string|decimal).
+func (b *builtinVitessHashSig) evalInt(row chunk.Row) (int64, bool, error) {
+	argTp := b.args[0].GetType().EvalType()
+	switch argTp {
+	case types.ETDecimal:
+		shardKeyDec, isNull, err := b.args[0].EvalDecimal(b.ctx, row)
+		if isNull || err != nil {
+			return 0, true, err
+		}
+		shardKeyUint, err := shardKeyDec.ToUint()
```

Comment:
seems we can require the first argument to `int`, no need to manually add a `cast` here?

Reply:
Yeah, removed all the casts.